java-checks/src/main/java/org/sonar/java/checks/LocksNotUnlockedCheck.java

```diff
@@ -0,0 +1,537 @@
+      scan(tree);
+      executionState = executionState.restoreParent();
+    }
+  }
+
+  private static class LockedOccurence {
+
+    @Nullable
+    private Tree lastAssignment;
+    private State state;
+
+    public LockedOccurence(@Nullable Tree lastAssignment, State state) {
+      this.lastAssignment = lastAssignment;
+      this.state = state;
+    }
+
+    @Override
+    public String toString() {
+      return ((JavaTree) lastAssignment).getLine()+" : "+state.name();
+    }
+  }
+
+  private static class ExecutionState {
+    @Nullable
+    private ExecutionState parent;
+    private Map<Symbol, LockedOccurence> lockedOccurencesBySymbol = Maps.newHashMap();
+    private IssuableSubscriptionVisitor check;
+
+    ExecutionState(IssuableSubscriptionVisitor check) {
+      this.check = check;
+    }
+
+    public ExecutionState(ExecutionState parent) {
+      this.parent = parent;
+      this.check = parent.check;
+    }
+
+    @Override
+    public String toString() {
+      String s = "";
+      for (Entry<Symbol, LockedOccurence> symbolLockedOccurenceEntry : lockedOccurencesBySymbol.entrySet()) {
+        LockedOccurence value = symbolLockedOccurenceEntry.getValue();
+        s += "  " + symbolLockedOccurenceEntry.getKey().name() + "  " + ((JavaTree) value.lastAssignment).getLine() + "  " + value.state.name() + "  ";
+      }
+      if(s.isEmpty()) {
+        s += " no locked occurences ";
+      }
+      return s;
+    }
+
+    public ExecutionState merge(ExecutionState executionState) {
+      for (Entry<Symbol, LockedOccurence> entry : executionState.lockedOccurencesBySymbol.entrySet()) {
+        Symbol symbol = entry.getKey();
+        LockedOccurence currentOccurence = getLockedOccurence(symbol);
+        LockedOccurence occurenceToMerge = entry.getValue();
+        if (currentOccurence != null) {
+          currentOccurence.state = currentOccurence.state.merge(occurenceToMerge.state);
+          lockedOccurencesBySymbol.put(symbol, currentOccurence);
+        } else {
+          // possible way to solve the problem of variable defined in outer Execution state. lockedOccurencesBySymbol.put(symbol,
+          // occurenceToMerge);
+          if (occurenceToMerge.state.isLocked()) {
+            insertIssue(occurenceToMerge.lastAssignment);
+          }
+        }
+      }
+      return this;
+    }
+
+    public ExecutionState overrideBy(ExecutionState currentES) {
+      for (Entry<Symbol, LockedOccurence> entry : currentES.lockedOccurencesBySymbol.entrySet()) {
+        Symbol symbol = entry.getKey();
+        LockedOccurence occurence = entry.getValue();
+        if (getLockedOccurence(symbol) != null) {
+          markAs(symbol, occurence.state);
+        } else {
+          lockedOccurencesBySymbol.put(symbol, occurence);
+        }
+      }
+      return this;
+    }
+
+    public ExecutionState restoreParent() {
+      if (parent != null) {
+        // insertIssues();
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)This block of commented-out lines of code should be removed.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)